networkx/linalg/laplacianmatrix.py

```diff
@@ -147,36 +147,117 @@ def normalized_laplacian_matrix(G, nodelist=None, weight="weight"):
+        Directed graphs:
+            The sum of the total multiplicative weights for all spanning trees of `G`,
+            rooted at node `root`.
+
+    Raises
+    ------
+    NetworkXPointlessConcept
+        If `G` does not contain any nodes.
+
+    NetworkXError
+        If the graph `G` is not (weakly) connected,
+        or if `G` is directed and the root node is not specified or not in G.
+
+    Examples
+    --------
+    >>> G = nx.complete_graph(5)
+    >>> round(nx.total_spanning_tree_weight(G))
+    125
+
+    >>> G = nx.Graph()
+    >>> G.add_edge(1, 2, weight=2)
+    >>> G.add_edge(1, 3, weight=1)
+    >>> G.add_edge(2, 3, weight=1)
+    >>> round(nx.total_spanning_tree_weight(G, "weight"))
+    5
+
+    Notes
+    -----
+    Self-loops are excluded. Multi-edges are contracted in one edge
+    equal to the sum of the weights.
+
+    References
+    ----------
+    .. [1] Wikipedia
+       "Kirchhoff's theorem."
+       https://en.wikipedia.org/wiki/Kirchhoff%27s_theorem
+    .. [2] Kirchhoff, G. R.
+        Über die Auflösung der Gleichungen, auf welche man
+        bei der Untersuchung der linearen Vertheilung
+        Galvanischer Ströme geführt wird
+        Annalen der Physik und Chemie, vol. 72, pp. 497-508, 1847.
+    .. [3] Margoliash, J.
+        "Matrix-Tree Theorem for Directed Graphs"
+        https://www.math.uchicago.edu/~may/VIGRE/VIGRE2010/REUPapers/Margoliash.pdf
     """
     import numpy as np
+    import scipy as sp
+
+    graph_is_directed = nx.is_directed(G)
+    if not graph_is_directed and not nx.is_connected(G):
+        raise nx.NetworkXError("Graph G must be connected.")
+    if graph_is_directed and root is None:
+        raise nx.NetworkXError("Spanning trees in directed graphs require a root node.")
+    if graph_is_directed and root not in G:
+        raise nx.NetworkXError("The node root is not in the graph G.")
+    if graph_is_directed and not nx.is_weakly_connected(G):
+        raise nx.NetworkXError("Graph G must be weakly connected.")
+
+    # Compute directed Laplacian matrix
+    if graph_is_directed == False:
+        L = nx.laplacian_matrix(G, weight=weight)
+    else:
+        A = nx.adjacency_matrix(G, weight=weight)
+        out_deg = [d for a, d in G.out_degree(weight=weight)]
+        n, m = A.shape
+        # TODO: rm csr_array wrapper when spdiags can produce arrays
+        D = sp.sparse.csr_array(sp.sparse.spdiags(out_deg, 0, m, n, format="csr"))
+        L = D - A
+
+    # Remove one row/column corresponding to node `root`
+    if graph_is_directed:
+        i = list(G).index(root)
+    else:
+        i = 0
+    L = L.todense()
+    L = np.delete(L, i, axis=0)
+    L = np.delete(L, i, axis=1)
 
-    G_laplacian = nx.laplacian_matrix(G, weight=weight).toarray()
-    # Determinant ignoring first row and column
-    return abs(np.linalg.det(G_laplacian[1:, 1:]))
+    # Compute sum of weight of all spanning trees
+    return np.linalg.det(L)
```

Comment:
The original implementation did not allow for weighted graphs, thus the absolute value was allowed. 
For graphs with negative weights, the number of spanning trees can be negative, see test `test_tstw_negative_weight`

Reply:
One of the important properties of the laplacian matrix is that it is positive semi-definite. When you consider negative weights the laplacian matrix is no longer positive semi-definite.  
xLx^T  = sum_{(i,j)\in E} w_{ij} * (x_i-x_j)^2  and this is positive for positive weights (and real vectors x).

The reason we use the `abs` is to remove the possibility of a negative result due to round-off errors. If there are such negative results due to round-off, they will be small so we reasoned that taking the `abs` will introduce an error of the size of round-off error. And the result will be positive.

BUT...  I don't know if the Kirchhoff theorem relies on the positive semi-definite feature of Laplacians. I know that there can't be a negative number of spanning trees. :)  But your test shows that with negative weights there can be negative cofactors of the laplacian. I guess the question is whether the theorem depends on the weights being positive or not. Do you have some insight into that question?

Thanks...

Reply:
@dschult I understand the reasoning for using `abs`, but it is generally not correct. Negative weights are allowed, because the theorem does not make use of the positive definiteness of the Laplacian matrix. Of course, negative weights are not commonly used, but the generalisation is straightforward and I couldn't find any counter-examples where the number of spanning trees suddenly becomes negative for non-negative weighted graphs (due to rounding errors) so I thought this was a good solution. 

Happy to hear your take on this.